src/wp-includes/html-api/class-wp-html-processor.php

```diff
@@ -1050,7 +1050,211 @@ private function step_before_head(): bool {
+		$token_type = $this->get_token_type();
+		$is_closer  = parent::is_tag_closer();
+		$op_sigil   = '#tag' === $token_type ? ( $is_closer ? '-' : '+' ) : '';
+		$op         = "{$op_sigil}{$token_name}";
+
+		/*
+		 * > A character token that is one of U+0009 CHARACTER TABULATION,
+		 * > U+000A LINE FEED (LF), U+000C FORM FEED (FF),
+		 * > U+000D CARRIAGE RETURN (CR), or U+0020 SPACE
+		 */
+		if ( '#text' === $op ) {
+			$text = $this->get_modifiable_text();
+			if ( '' === $text ) {
+				/*
+				 * If the text is empty after processing HTML entities and stripping
+				 * U+0000 NULL bytes then ignore the token.
+				 */
+				return $this->step();
+			}
+
+			if ( strlen( $text ) === strspn( $text, " \t\n\f\r" ) ) {
+				// Insert the character.
+				$this->insert_html_element( $this->state->current_token );
+				return true;
+			}
+		}
+
+		switch ( $op ) {
+			/*
+			 * > A comment token
+			 */
+			case '#comment':
+			case '#funky-comment':
+			case '#presumptuous-tag':
+				$this->insert_html_element( $this->state->current_token );
+				return true;
+
+			/*
+			 * > A DOCTYPE token
+			 */
+			case 'html':
+				// Parse error: ignore the token.
+				return $this->step();
+
+			/*
+			 * > A start tag whose tag name is "html"
+			 */
+			case '+HTML':
+				return $this->step_in_body();
+
+			/*
+			 * > A start tag whose tag name is one of: "base", "basefont", "bgsound", "link"
+			 */
+			case '+BASE':
+			case '+BASEFONT':
+			case '+BGSOUND':
+			case '+LINK':
+				$this->insert_html_element( $this->state->current_token );
+				return true;
+
+			/*
+			 * > A start tag whose tag name is "meta"
+			 */
+			case '+META':
+				$this->insert_html_element( $this->state->current_token );
+
+				/*
+				 * > If the active speculative HTML parser is null, then:
+				 * >   - If the element has a charset attribute, and getting an encoding from
+				 * >     its value results in an encoding, and the confidence is currently
+				 * >     tentative, then change the encoding to the resulting encoding.
+				 */
+				$charset = $this->get_attribute( 'charset' );
+				if ( is_string( $charset ) ) {
+					$this->bail( 'Cannot yet process META tags with charset to determine encoding.' );
+				}
+
+				/*
+				 * >   - Otherwise, if the element has an http-equiv attribute whose value is
+				 * >     an ASCII case-insensitive match for the string "Content-Type", and
+				 * >     the element has a content attribute, and applying the algorithm for
+				 * >     extracting a character encoding from a meta element to that attribute's
+				 * >     value returns an encoding, and the confidence is currently tentative,
+				 * >     then change the encoding to the extracted encoding.
+				 */
+				$http_equiv = $this->get_attribute( 'http-equiv' );
+				$content    = $this->get_attribute( 'content' );
+				if (
+					is_string( $http_equiv ) &&
+					is_string( $content ) &&
+					0 === strcasecmp( $http_equiv, 'Content-Type' )
```

Comment:
Ditto above. Shouldn't this only bail if the charset is not UTF-8?

I guess this `bail` is essentially a TODO.